### ISSUE TITLE
Metadata extraction does not seem to follow the spec

### ISSUE THREAD

**Issue by RubenVerborgh** [comunica/comunica#121]
#### Issue type:

- :bug: Bug

#### Description:

It seems that #15 does not fully follow the TPF spec at https://www.hydra-cg.com/spec/latest/triple-pattern-fragments/.

What should happen is that the server looks for
```?dataset void:subset <CURRENTLOCATION>;
    hydra:search [
        hydra:template ?template;
        hydra:mapping  [ hydra:variable ?s; hydra:property rdf:subject ],
                       [ hydra:variable ?p; hydra:property rdf:predicate ],
                       [ hydra:variable ?o; hydra:property rdf:object ]
    ].
```

IF a quad-based format, then that pattern MUST be matched in a named graph
and that named graph SHOULD be related to the document.

Note that foaf:primaryTopic doesn’t play any role in this, except for SHOULD.

**Comment by RubenVerborgh** [comment]
Bonus: implement this as a SPARQL query rather than as a hard-coded lookup algorithm.

**Comment by rubensworks** [comment]
Ok, so `actor-rdf-metadata-primary-topic` is indeed too strict.
This can easily be fixed by removing that actor from the default config.
(*I won't do this yet until I can get a second opinion on this*)

We should however make a smarter mediator in the future that can still make use of that actor _if_ `foaf:primaryTopic` is present. And we should probably also add another actor that does the named graph match thingy as mentioned above.

However, @pietercolpaert's query still fails for me even if I disable this actor:
```
https://linked.open.gent/parking "SELECT * WHERE { ?s ?p ?o }"
```

I did some debugging, and it looks like the metadata of https://linked.open.gent/parking contains the following:
```
<https://linked.open.gent/parking#mappingS> hydra:variable "s".
<https://linked.open.gent/parking#mappingP> hydra:variable "p".
<https://linked.open.gent/parking#mappingO> hydra:variable "o".
<https://linked.open.gent/parking#mappingS> hydra:property "subject".
<https://linked.open.gent/parking#mappingP> hydra:property "property".
<https://linked.open.gent/parking#mappingO> hydra:property "object".
```
This is however incorrect, as `hydra:property` has a URI as range AFAIK, not a string, so this should be:
```
<https://linked.open.gent/parking#mappingS> hydra:variable "s".
<https://linked.open.gent/parking#mappingP> hydra:variable "p".
<https://linked.open.gent/parking#mappingO> hydra:variable "o".
<https://linked.open.gent/parking#mappingS> hydra:property rdf:subject.
<https://linked.open.gent/parking#mappingP> hydra:property rdf:property.
<https://linked.open.gent/parking#mappingO> hydra:property rdf:object.
```

These forms were already being detected by comunica, it just didn't them as _valid_ due to the missing `rdf:subject` etc., links. So changing this should fix it :-)

**Comment by pietercolpaert** [comment]
Closed by accident, yet I did fix the wrong metadata template in the parking data pilot. This is now corrected. Yet still the right metadata is not extracted due to the `void:subset` predicate not being followed.

**Comment by rubensworks** [comment]
The SPARQL query-based metadata lookup will be followed up in https://github.com/comunica/comunica/issues/115

**Comment by rubensworks** [comment]
Actually, the latest commit causes issues with QPF.

@RubenVerborgh As far as I can see, there is no spec-compliant way to determine the metadata graph when using QPF. What do you think about changing the spec to say?:
_IF a quad-based format, then that pattern MUST be matched in a named graph
and that named graph MUST be related to the document using `foaf:primaryTopic`._

**Comment by rubensworks** [comment]
Alright, the primary topic metadata actor has now been enabled, and will always be used for quad serializations. This means that comunica _requires_ the metadata graph to be linked to the document using `foaf:primaryTopic`.

@RubenVerborgh If you have any other ideas on this that does not require changing the spec, please let me know. (Special-casing QPF would be possible, but that feels like a hack)

**Comment by pietercolpaert** [comment]
Did the spec already change to do this? I’m a bit confused what is requested now. Should it be like this?

HTTP GET <URI1>
```trig
<#metadata> {
   <URI1> a foaf:Document ;
                foaf:primaryTopic <#metadata> .
}
```

Or should it be the reverse?

**Comment by rubensworks** [comment]
Comunica is indeed expecting the reverse when a quad-format is used. The spec has not been changed with respect to this, still waiting on @RubenVerborgh for his view on that.

**Comment by RubenVerborgh** [comment]
Proposal: Comunica picks up all TPF/QPF forms, regardless of the graph. Would that make sense?

**Comment by RubenVerborgh** [comment]
@pietercolpaert Well, any response can contain multiple Hydra and thus TPF forms. We are specifically looking for the form where, given a current fragment URL of <u>:

```
?dataset void:subset <u>;
    hydra:search ?form.
```

**Comment by rubensworks** [comment]
@RubenVerborgh Picking up all forms in all graphs does make sense. But that doesn't solve the problem of distinguishing metadata (which we would still have) from data.

For example, the following metadata is also emitted by TPF in TriG:
```
<https://fragments.linkedsoftwaredependencies.org/npm> void:subset <https://fragments.linkedsoftwaredependencies.org/npm>;
    a hydra:PartialCollectionView;
    dc:title "Linked Data Fragment of NPM"@en;
    dc:description "Triple Pattern Fragment of the 'NPM' dataset containing triples matching the pattern { ?s ?p ?o }."@en;
    dc:source <https://fragments.linkedsoftwaredependencies.org/npm#dataset>;
    hydra:totalItems "328924262"^^xsd:integer;
    void:triples "328924262"^^xsd:integer;
    hydra:itemsPerPage "100"^^xsd:integer;
    hydra:first <https://fragments.linkedsoftwaredependencies.org/npm?page=1>;
    hydra:next <https://fragments.linkedsoftwaredependencies.org/npm?page=2>
}
```

Any metadata can be defined in this graph, so unless we have a way of properly determining the metadata graph, we can't exclude these triples from the data stream.

**Comment by pietercolpaert** [comment]
I think I understand what @RubenVerborgh wants to say with _should_: if you do so, these metadata triples will be excluded from the result, if no `<#metadata> foaf:primaryTopic <documenturi>` is present, then also the metadata are part of the result, which will not give you any real problems?

**Comment by RubenVerborgh** [comment]
> But that doesn't solve the problem of distinguishing metadata (which we would still have) from data.

Right, that's a very good point.

However, there are two schools of thought. One of them is that the metadata graph doesn't belong in the query results, the other is that it does. You can argue both ways (so maybe we should make it configurable). The metadata graph is an interesting side-effect of TPF, but if we consider the document as a Linked Data document, then the metadata graphs are there.

Moreover, a 100% clean separation doesn't exist. Datasets could, in theory, also include `foaf:primaryTopic` relations that exactly look like the ones we use.

And there's also @pietercolpaert's point above, which I really like.

So given these thoughts, what do you think?

**Comment by rubensworks** [comment]
> However, there are two schools of thought. One of them is that the metadata graph doesn't belong in the query results, the other is that it does.

I agree with @pietercolpaert's comment that the publisher should be able to decide this using `foaf:primaryTopic`.

If a `foaf:primaryTopic` link is present, then the distinction between data and metadata is (mostly) clear. But what should happen if this is not present? Should everything be seen as both data and metadata? Or should everything that _could_ be seen as metadata, seen as metadata, and excluded from the data?

More concretely, in non-quad serializations, we now assume all Hydra-based triples to be seen as part of the metadata, and _not_ the data. Is this the "correct" way to do this, or should these triples also be made part of the data? Or is this exactly the part that should be made configurable?

Related to this, should we make metadata distinction in quad serializations fall back to this non-quad case (Hydra-based triples) when there is no proper `foaf:primaryTopic` link present?

**Comment by RubenVerborgh** [comment]
> I agree with @pietercolpaert's comment that the publisher should be able to decide this using foaf:primaryTopic.

Okay, then we should probably make a note in the TPF spec that the presence of this predicate allows the client the remove the graph from the output.

> If a foaf:primaryTopic link is present, then the distinction between data and metadata is (mostly) clear. But what should happen if this is not present? Should everything be seen as both data and metadata?

Yes, as far as I'm concerned.

> More concretely, in non-quad serializations, we now assume all Hydra-based triples to be seen as part of the metadata, and not the data. Is this the "correct" way to do this, or should these triples also be made part of the data?

That's a heuristic, which I would drop. I'd propose:
- if `foaf:primaryTopic` with an object of `responseURL` is present, then consider this graph metadata, drop it as data
- else, consider everything as both data and metadata

> Or is this exactly the part that should be made configurable?

The configuration for me would be to not drop metadata, even with `foaf:primaryTopic`.

> Related to this, should we make metadata distinction in quad serializations fall back to this non-quad case (Hydra-based triples) when there is no proper foaf:primaryTopic link present?

Indeed, that's what it comes down to. So rephrased: if there is an explicit indication of metadata graph, we give it a special treatment. If not, we look for metadata everywhere.

**Comment by pietercolpaert** [comment]
I think making it configurable is a bad idea as it would leave the door open for, in the long run, issues that will appear because “for that server you should configure your client differently”.

**Comment by RubenVerborgh** [comment]
> “for that server you should configure your client differently”

Mmm no, it's more a question of interpretation really, which is purely the client. Do you want to interpret the metadata graph as data or not?
The latter could be useful in cases where, for instance, you want to make a study of hypermedia forms on the Web.

**Comment by rubensworks** [comment]
> That's a heuristic, which I would drop. I'd propose:
> * if foaf:primaryTopic with an object of responseURL is present, then consider only this graph as metadata, and drop it as data
> * else, consider everything as both data and metadata

Not sure if we should drop the current heuristic. This would break backwards compatibility with triple-only serializations (as is the case with TPF Java servers). In those cases, queries such as `?s ?p ?o` would suddenly also start including metadata result. This is probably not something we want, or do we?

> The configuration for me would be to not drop metadata, even with `foaf:primaryTopic`.

The default behaviour would be to always exclude metadata from data (when possible) I guess.

**Comment by RubenVerborgh** [comment]
> Not sure if we should drop the current heuristic. This would break backwards compatibility with triple-only serializations (as is the case with TPF Java servers). In those cases, queries such as ?s ?p ?o would suddenly also start including metadata result.

Well, the separation was never perfect anyway.
TPF in the beginning only had triple serializations (historical note: I actually extended N3.js to quad formats for TPF). And some metadata triples like `dc:title` would show up in the result with `?s ?p ?o` queries. As argued above, this can be interpreted as being correct, since it _is_ part of the data.

> This is probably not something we want, or do we?

I think it is acceptable, and in some interpretations even correct.
The breakage is limited to returning extra triples, which is not that bad (and we are already returning some extra triples anyway in these cases).

> The default behaviour would be to always exclude metadata from data (when possible) I guess.

Indeed.

**Comment by rubensworks** [comment]
> I think it is acceptable, and in some interpretations even correct.

Ok, just wanted to double-check that :-)

So in summary, this is the plan:
* Remove heuristical Hydra-based metadata identification.
* If `foaf:primaryTopic` is present (with proper link to current resource), make a concrete distinction between `data` and `metadata` streams.
* Optionally, the `metadata` stream can be merged into the `data` stream through a config option
* If `foaf:primaryTopic` is _not_ present, everything is seen as `data` and no `metadata` stream exists.
* Modify metadata extractors (hydra forms and count) so that if `metadata` is not present, they will look for the information in the `data` stream. If `metadata` _is_ present, they will look for the information in there, which _may_ lead to better performance if the extractors are complex.

Let's make this part of the 2.0.0 milestone.

**Comment by RubenVerborgh** [comment]
Yes, but I wouldn't add that knowledge to the extractors. Something should just give them a stream, no?

**Comment by rubensworks** [comment]
Sure, we could abstract this into the actors that are responsible for creating the metadata stream.

**Comment by RubenVerborgh** [comment]
Concrete implementation idea: if config is set, filter the data stream. If not set, filter a clone so all triples are preserved.

**Comment by pietercolpaert** [comment]
@larsgsvensson `<#metadata> foaf:primaryTopic <current url>`

**Comment by RubenVerborgh** [comment]
... And that triple only makes sense in quad-based representations, where there is a #metadata graph.

**Comment by larsgsvensson** [comment]
Thanks @pietercolpaert,

 

so just to be sure, if I call `http://fragments.dbpedia.org/2015/en?subject=&predicate=&object=http%3A%2F%2Fdbpedia.org%2Fresource%2FBelgium` the returned data should contain the triple

```turtle

<http://fragments.dbpedia.org/2015/en?subject= <http://fragments.dbpedia.org/2015/en?subject=&predicate=&object=http%3A%2F%2Fdbpedia.org%2Fresource%2FBelgium#metadata> &predicate=&object=http%3A%2F%2Fdbpedia.org%2Fresource%2FBelgium#metadata> foaf:primaryTopic http://fragments.dbpedia.org/2015/en?subject= <http://fragments.dbpedia.org/2015/en?subject=&predicate=&object=http%3A%2F%2Fdbpedia.org%2Fresource%2FBelgium> &predicate=&object=http%3A%2F%2Fdbpedia.org%2Fresource%2FBelgium

```

Is that correct?

**Comment by larsgsvensson** [comment]
@RubenVerborgh scripsit
> ... And that triple only makes sense in quad-based representations, where there is a #metadata graph.

So which triple would make sense?

**Comment by RubenVerborgh** [comment]
> Is that correct?

Yes.

> So which triple would make sense?

In a triple-based representation, there is no separate metadata graph, so no `foaf:primaryTopic` triple is necessary. The purpose of this triple is to, in  quad-based representations, indicate the metadata graph.

**Comment by larsgsvensson** [comment]
> In a triple-based representation, there is no separate metadata graph, so no `foaf:primaryTopic` triple is necessary. The purpose of this triple is to, in quad-based representations, indicate the metadata graph.

But then how does the client filter out the metadata triples? I currently experiment using comunica with a triple-based implementation and when I run the query
```sparql
PREFIX gndo: <http://d-nb.info/standards/elementset/gnd#>
SELECT ?s 
WHERE { ?s a gndo:DifferentiatedPerson }
```
`?s` also contains the subject of the metata (in this case `http://10.69.14.96:8081/authorities#dataset`) so there must be some way to filter those triples (as there must be some way to filter the extra data triples that don't match the pattern but the server returns anyway).

**Comment by rubensworks** [comment]
@larsgsvensson So I assume your (triple-based) endpoint returns the triple `<http://10.69.14.96:8081/authorities#dataset> a void:Dataset`?

If so, then the results of this query should definitely not contain `?s = <http://10.69.14.96:8081/authorities#dataset>`. Could you open up a new issue for this? As this sounds like a different bug somewhere in the client-side filtering.

**Comment by RubenVerborgh** [comment]
> But then how does the client filter out the metadata triples?

Using a heuristic (i.e., drop all Hydra predicates). With triple-based representations, a perfect separation of data and metadata is not possible.

**Comment by larsgsvensson** [comment]
@rubensworks 
>  So I assume your (triple-based) endpoint returns the triple `<http://10.69.14.96:8081/authorities#dataset> a void:Dataset?`

Yes, that's correct. Here are all triples returned with subject `<http://10.69.14.96:8081/authorities#dataset>`:
```turtle
<http://10.69.14.96:8081/authorities#dataset>
    a void:Dataset , hydra:Collection ;
    void:subset <http://10.69.14.96:8081/authorities?predicate=http%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23type&object=http%3A%2F%2Fd-nb.info%2Fstandards%2Felementset%2Fgnd%23DifferentiatedPerson> ;
    hydra:itemsPerPage  "100"^^xsd:long ;
    hydra:search [
        hydra:mapping
            [ hydra:property  rdf:subject ;
              hydra:variable  "subject"
            ] ;
        hydra:mapping
            [ hydra:property  rdf:predicate ;
              hydra:variable  "predicate"
            ] ;
         hydra:mapping
             [ hydra:property  rdf:object ;
               hydra:variable  "object"
             ] ;
          hydra:template  "http://10.69.14.96:8081/authorities{?subject,predicate,object}"
      ] .
```
I've opened #152 for this.

**Comment by larsgsvensson** [comment]
@RubenVerborgh 
> Using a heuristic (i.e., drop all Hydra predicates). With triple-based representations, a perfect separation of data and metadata is not possible.

Of course it isn't but since the server is allowed to return a superset of what the client asked for, the client must filter all triples it receives anyway in order to see if they really match the required pattern. Or do I misunderstand something?

**Comment by RubenVerborgh** [comment]
@larsgsvensson No, you're right. But that means that some queries, such as`?s ?p ?o`, will give too many results with triple-based representations.

**Comment by larsgsvensson** [comment]
@RubenVerborgh But can't the client filter the received triples and throw away those that don't match the pattern it asked for?

**Comment by RubenVerborgh** [comment]
> @RubenVerborgh But can't the client filter the received triples and throw away those that don't match the pattern it asked for?

Depending on the pattern, yes. E.g., if you ask a TPF for `dbpedia:Nikola_Tesla ?p ?o`, then a client will be able to get rid of all metadata. More difficult with a TPF for `?s ?p ?o` or `?s foaf:primaryTopic ?o`.

> (And asking for ?s ?p ?o is something you shouldn't do anyway...)

It depends; if the client is given the task of solving `SELECT DISTINCT ?o { ?s ?p ?o FILTER(LANGMATCHES(LANG(?o), 'EN')) }`, then there's not really an alternative with TPF.

**Comment by larsgsvensson** [comment]
@RubenVerborgh 
OK, I see your point. Do we agree that it would make sense that the client throws away those triples that don't match the requested pattern where possible? (And that we document somewhere which patterns the client won't be able to filter?)

**Comment by RubenVerborgh** [comment]
> Do we agree that it would make sense that the client throws away those triples that don't match the requested pattern where possible?

Absolutely.

> (And that we document somewhere which patterns the client won't be able to filter?)

That's tricky, as it depends on what additional triples the server sends. And as you remarked above, the server is free to send any additional triples. So the answer, in general, is “metadata triples that also match the data selector”.

**Comment by rubensworks** [comment]
Fixed in https://github.com/comunica/comunica/commits/refactor/metadata, will close this issue once this branch has been merged.
